src/Google.Api.Gax/Api/Gax/GetOrCreateHelper.cs

```diff
@@ -0,0 +1,153 @@
+        /// resource doesn't already exist.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// If <paramref name="fetcher"/> fails with an <see cref="RpcException"/> with a status code
+        /// of <see cref="StatusCode.NotFound"/>, it is assumed that the resource needs to be created, and
+        /// <paramref name="creator"/> is called. If this then fails with an <see cref="RpcException"/> with
+        /// a status code of <see cref="StatusCode.AlreadyExists"/>, it is assumed that two clients have been
+        /// racing to create the resource, and the whole process is retried. The maximum number of attempts
+        /// (whole get/create cycles) is limited by <paramref name="maxAttempts"/>.
+        /// </para>
+        /// <para>
+        /// It is expected that if cancellation support is required, it will be embedded in the calls to <paramref name="fetcher"/>
+        /// and <paramref name="creator"/>.
+        /// </para>
+        /// <para>
+        /// If <paramref name="validator"/> is not <c>null</c>, it is called when a resource is fetched rather than created,
+        /// with the fetched resource being passed in. If the result of the validation is non-null, it is used as the message
+        /// of a <see cref="ResourceMismatchException"/> which is thrown as the result of the operation.
+        /// </para>
+        /// </remarks>
+        /// <typeparam name="TResource">The type of the resource being fetched or created.</typeparam>
+        /// <param name="fetcher">A delegate to asynchronously fetch the resource. Must not be null.</param>
+        /// <param name="creator">A delegate to asynchronously create the resource if it doesn't already exist. Must not be null.</param>
+        /// <param name="validator">An optional delegate to validate that a fetched resource has the expected form.</param>
+        /// <param name="maxAttempts">The maximum number of fetch/create cycles to try before giving up.</param>
+        /// <exception cref="ResourceMismatchException">The resource was fetched, but failed the specified validation.</exception>
+        /// <returns>A task representing the asynchronous operation. When the task is completed, its result will be
+        /// the resource fetched or created.</returns>
+        public static async Task<TResource> GetOrCreateAsync<TResource>(
+            Func<Task<TResource>> fetcher,
+            Func<Task<TResource>> creator,
+            Func<TResource, string> validator = null,
+            int maxAttempts = DefaultMaxAttempts)
+        {
+            GaxPreconditions.CheckNotNull(fetcher, nameof(fetcher));
+            GaxPreconditions.CheckNotNull(creator, nameof(creator));
+            GaxPreconditions.CheckArgumentRange(maxAttempts, nameof(maxAttempts), 1, int.MaxValue);
+            for (int i = 0; i < maxAttempts; i++)
+            {
+                try
+                {
+                    var resource = await fetcher().ConfigureAwait(false);
```

Comment:
Since you are calling user provided callbacks here, we should either document the fact that they can be called on different threads and/or don't user `ConfigureAwait` to ensure that they are called in the same context.